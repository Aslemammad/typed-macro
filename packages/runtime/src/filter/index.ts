export { FilterOptions, FilterPattern, createFilter, Filter } from './filter'
