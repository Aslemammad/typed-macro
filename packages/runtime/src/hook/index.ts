export { createHookContainer, HookContainer, CombinedHooks } from './hook'
