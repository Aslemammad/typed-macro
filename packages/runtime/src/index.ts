export * from './transformer'
export * from './normalizer'
export * from './renderer'
export * from './runtime'
export * from './loader'
export * from './filter'
export * from './hook'
