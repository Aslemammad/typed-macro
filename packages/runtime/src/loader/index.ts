export { Loader, createLoader } from './loader'
