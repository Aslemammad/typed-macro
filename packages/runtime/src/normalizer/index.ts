export {
  ModularizedMacros,
  ModularizedModules,
  ModularizedTypes,
  NormalizedExports,
  normalizeExports,
  normalizeMacroProvider,
  NormalizedMacroProvider,
} from './normalizer'
