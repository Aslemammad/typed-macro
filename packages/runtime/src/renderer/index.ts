export { TypeRenderer, createTypeRenderer } from './renderer'
