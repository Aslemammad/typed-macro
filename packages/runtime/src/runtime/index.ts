export { Runtime, RuntimeOptions, createRuntime } from './runtime'
