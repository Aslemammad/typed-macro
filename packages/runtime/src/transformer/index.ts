export {
  Transformer,
  createTransformer,
  TransformerOptions,
} from './transformer'
