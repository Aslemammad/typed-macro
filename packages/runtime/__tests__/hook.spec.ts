import { createHookContainer } from '../src'
import { MacroProviderHooks } from '@typed-macro/core'

describe('HookContainer', () => {
  it('should work', async () => {
    const hooks = createHookContainer()

    const executed: string[] = []

    const mockedHooksA = {
      _: () => {
        executed.push('hooks A: bad hook!')
      },
      onStart: () => {
        executed.push('hooks A: onStart()')
      },
      onStop: () => {
        executed.push('hooks A: onStop()')
      },
      beforeTransform: (code: string) => {
        executed.push('hooks A: beforeTransform()')
        return 'a' + code
      },
      afterTransform: () => {
        executed.push('hooks A: afterTransform()')
        return {}
      },
    } as any

    const mockedHooksB: MacroProviderHooks = {
      onStart: () => {
        executed.push('hooks B: onStart()')
      },
      onStop: () => {
        executed.push('hooks B: onStop()')
      },
      beforeTransform: (code) => {
        executed.push('hooks B: beforeTransform()')
        return {
          code: 'b' + code,
        }
      },
      afterTransform: (code) => {
        executed.push('hooks B: afterTransform()')
        return {
          code: code + 'b',
        }
      },
    }
    hooks.append(mockedHooksA)
    hooks.append(mockedHooksB)

    await hooks.onStart()
    await hooks.onStop()
    expect(await hooks.beforeTransform('code', '')).toMatchSnapshot()
    expect(await hooks.afterTransform('code', '')).toMatchSnapshot()

    expect(executed).toMatchSnapshot()
  })

  it('should have short circuit rule in filter()', async () => {
    const filterTrue = jest.fn(() => true)
    const filterFalse = jest.fn(() => false)

    const hooksA = createHookContainer()
    hooksA.append({ onFilter: filterFalse })
    hooksA.append({ onFilter: filterTrue })
    hooksA.append({ onFilter: filterTrue })
    expect(await hooksA.onFilter('')).toBe(true)
    expect(filterTrue.mock.calls.length).toBe(1)
    expect(filterFalse.mock.calls.length).toBe(1)

    const hooksB = createHookContainer()
    hooksB.append({ onFilter: filterFalse })
    hooksB.append({ onFilter: filterFalse })
    hooksB.append({ onFilter: filterFalse })
    expect(await hooksB.onFilter('')).toBe(false)
    expect(filterTrue.mock.calls.length).toBe(1)
    expect(filterFalse.mock.calls.length).toBe(4)
  })
})
