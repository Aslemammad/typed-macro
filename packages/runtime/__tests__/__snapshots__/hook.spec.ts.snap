// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HookContainer should work 1`] = `"bacode"`;

exports[`HookContainer should work 2`] = `"codeb"`;

exports[`HookContainer should work 3`] = `
Array [
  "hooks A: onStart()",
  "hooks B: onStart()",
  "hooks A: onStop()",
  "hooks B: onStop()",
  "hooks A: beforeTransform()",
  "hooks B: beforeTransform()",
  "hooks A: afterTransform()",
  "hooks B: afterTransform()",
]
`;
